lib/index.js

```diff
@@ -19,6 +89,57 @@ function handleTokenRequest({logger}) {
     }
   };
 }
-odule.exports = {
-  handleTokenRequest
+
+function _authenticateClient({client, clientSecret}) {
+  if(!client) {
+    throw new InvalidClient({
+      description: 'Unknown client identifier.'
+    });
+  }
+
+  const storedClientSecret = client.clientSecret || client.client_secret;
+
+  if(clientSecret !== storedClientSecret) {
```

Comment:
This needs to be a constant time comparison.

Reply:
What does that mean?

Reply:
It can't be simple string comparison because that will leak timing information that will enable an attacker to figure out client secrets. Instead we need to use a constant-time API call to do the comparison, i.e., no matter what value is passed in, the comparison will take the same amount of time (or, alternatively, it will appear random to an attacker). We don't want attackers gaining information by failed comparisons.

In this case, we really shouldn't be storing the client secret at all if we can avoid that. We should instead HMAC it. Then we can HMAC the value given to us for comparison and use [crypto.timingSafeEqual](https://nodejs.org/dist/latest-v6.x/docs/api/crypto.html#crypto_crypto_timingsafeequal_a_b) to compare the resulting buffers containing the signature values/digests.

Reply:
Got it, thanks. Will open an issue for it.